File: downloads/code/LeetCode/Java/Linked-List-Random-Node.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {

    /** @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node. */
    ListNode head = null;
    Random randomGenerator = null;
    public Solution(ListNode head) {
        this.head = head;
        this.randomGenerator = new Random();
    }
    
    /** Returns a random node's value. */
    public int getRandom() {
        int ret=0;
        ListNode cur = this.head;
        for(int i=1;cur!=null;i++){
            if(randomGenerator.nextInt(i)==0)
                ret = cur.val;
            cur = cur.next;
        }
    return ret;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */

File: downloads/code/LeetCode/Python/Linked-List-Random-Node.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):

    def __init__(self, head):
        """
        @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
        :type head: ListNode
        """
        self.head = head        

    def getRandom(self):
        """
        Returns a random node's value.
        :rtype: int
        """
        i = 0
        ans = 0
        head = self.head
        while head:
            # randint 是在[0,i]中生成随机数
            # roll到0才算数
            if random.randint(0,i) == 0:
                ans = head.val
            # head 一直在前进
            head = head.next
            i = i+
        return ans

# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()
